urle381b6cba7_IkarosKappler_ikrs_http_tgz-pJ8-ikrs_json_rpc_JSONRPCHandlerJ8/src/ikrs/httpd/resource/ReplacingResource.java

```diff
@@ -0,0 +1,134 @@
+     * @param useFairLocks   If set to true the class will use fair read locks (writing isn't
+     *                       possible at all with this class).
+     * @param replacementMap The map containing the (placeholder replacement) pairs; the map
+     *                       may be empty but must not be null.
+     * @throws NullPointerException If logger, the input stream, the replacement map is null
+     *                              or if the replacement map contains null entries.
+     **/
+    public ReplacingResource( HTTPHandler handler,
+			      CustomLogger logger,
+			      InputStream in,
+			      boolean useFairLocks,
+			      
+			      Map<byte[],byte[]> replacementMap ) 
+	throws NullPointerException {
+
+	super( handler, logger, new ReplacingInputStream(in, replacementMap), useFairLocks );
+	
+    }
+
+
+
+    /**
+     * For testing only.
+     **/
+    public static void main( String[] argv ) {
+
+	
+	String filename = "document_root/system/errors/GenericError.template.html";
+	CustomLogger logger = new ikrs.util.DefaultCustomLogger("test");
+
+	Map<byte[],byte[]> replacementMap = new java.util.TreeMap<byte[],byte[]>( new ikrs.util.ByteArrayComparator() );
+	try {
+
+	    replacementMap.put( new String("%{STATUS_CODE}%").getBytes(java.nio.charset.StandardCharsets.UTF_8.name()),
+				new String("500").getBytes(java.nio.charset.StandardCharsets.UTF_8.name()) 
+				);
+	    replacementMap.put( new String("%{REASON_PHRASE}%").getBytes(java.nio.charset.StandardCharsets.UTF_8.name()),
+				new String("Iternal Server Error").getBytes(java.nio.charset.StandardCharsets.UTF_8.name()) 
+				);
+	    replacementMap.put( new String("%{ERROR_MESSAGE}%").getBytes(java.nio.charset.StandardCharsets.UTF_8.name()),
+				new String("I encountered an internal server error.").getBytes(java.nio.charset.StandardCharsets.UTF_8.name()) 
+				);
+	    System.out.println( "Replacement map: " + replacementMap.toString() );
+
+	    
+	    System.out.println( "Creating file resource from " + filename );
+	    ikrs.httpd.Resource fileResource = new FileResource( null,  // no handler available
+									  logger,
+									  new java.io.File(filename),
+									  false // no need to use fair locks here
+									  );
+	    System.out.println( "Opening file resource ... ");
+	    fileResource.open( true );  // open in read-only mode
+
+	    System.out.println( "Creating replacing resource ... ");
+	    ikrs.httpd.Resource replacingResource = new ReplacingResource( null,  
+									   logger,
+									   fileResource.getInputStream(),
+									   false,
+									   replacementMap
+									   );
+
+	    System.out.println( "Opening replacing resource ..." );
+	    replacingResource.open( true );
+
+	    System.out.println( "Reading replacing resource ... ");
+	    byte[] buffer = new byte[ 256 ];
+	    int len;
+	    while( (len = replacingResource.getInputStream().read(buffer)) != -1 ) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code contains a resource that might not be closed properly. This can cause a resource leak that slows down or crashes your system.

**Fix**
Consider closing the resource returned by the following method call: *getInputStream*. The resource is referenced at lines 96, 99, 101, 107, 112. The resource is closed at lines 121, 124. There are other execution paths that do not contain closure statements, for example, when InputStream.read throws an exception. Ensure the object returned by *getInputStream()* is closed in a try-finally block or declared in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).